nengo/networks/assoc_mem.py

```diff
@@ -0,0 +1,436 @@
+
+
+class AssociativeMemory(nengo.Network):
+    """Associative memory network.
+
+    Parameters
+    ----------
+    input_vectors: array_like
+        The list of vectors to be compared against.
+    output_vectors: array_like, optional
+        The list of vectors to be produced for each match. If
+        not given, the associative memory will act like an auto-associative
+        memory (cleanup memory).
+
+    n_neurons: int
+        The number of neurons for each of the ensemble (where each ensemble
+        represents each item in the input_vectors list)
+
+    threshold: array_like, optional
+        The association activation threshold.
+    input_scales: array_list, optional
+        Scaling factor to apply on each of the input vectors. Note that it
+        is possible to scale each vector independently.
+
+    inhibitable: boolean, optional
+        Flag to indicate if the entire associative memory module is
+        inhibitable (entire thing can be shut off).
+
+    threshold_output: boolean, optional
+        Flag to indicate if the output vector value should be thresholded.
+        By default, the shape of the output of the associative memory is a
+        filtered step function, shaped to reach the 0.8 * maximal output value
+        at threshold + 0.1, 0.95 * maximal output value at threshold + 0.2, and
+        maximal output value (1) at threshold + 0.3 (when the radius is 1).
+
+    """
+    def __init__(self, input_vectors, output_vectors=None,  # noqa: C901
+                 n_neurons=50, threshold=0.3, input_scales=1.0,
+                 inhibitable=False, config=None, label=None, seed=None,
+                 add_to_container=None):
+        super(AssociativeMemory, self).__init__(label, seed, add_to_container)
+
+        # Label prefix for all the ensemble labels
+        label_prefix = "" if label is None else label + "_"
+
+        # If output vocabulary is not specified, use input vector list
+        # (i.e autoassociative memory)
+        if output_vectors is None:
+            output_vectors = input_vectors
+
+        # Handle different vector list types
+        if is_iterable(input_vectors):
+            input_vectors = np.array(input_vectors, ndmin=2)
+
+        if is_iterable(output_vectors):
+            output_vectors = np.array(output_vectors, ndmin=2)
+
+        # Fail if number of input items and number of output items don't
+        # match
+        if input_vectors.shape[0] != output_vectors.shape[0]:
+            raise ValueError(
+                'Number of input vectors does not match number of output '
+                'vectors. %d != %d'
+                % (input_vectors.shape[0], output_vectors.shape[0]))
+
+        # Handle possible different threshold / input_scale values for each
+        # element in the associative memory
+        if not is_iterable(threshold):
+            threshold = np.array([threshold] * input_vectors.shape[0])
```

Comment:
Would it be better/more clear to write `threshold * np.ones(input_vectors.shape[0])`?